app/scripts/directives/editEnvironmentFrom.js

```diff
@@ -0,0 +1,204 @@
+"use strict";
+(function() {
+  angular.module("openshiftConsole").component('editEnvironmentFrom', {
+    controller: [
+      '$attrs',
+      '$filter',
+      '$scope',
+      'keyValueEditorUtils',
+      EditEnvironmentFrom
+    ],
+    controllerAs: 'ctrl',
+    //require: '^form',
+    bindings: {
+      addRowLink: '@',
+      entries: '=',
+      valueFromSelectorOptions: '<',
+      valuePlaceholder: '@'
+      //cannotAdd: '<?',
+      //cannotSort: '<?',
+      //cannotDelete: '<?',
+      //isReadonly: '<?',
+      //showHeader: '<?',
+    },
+    templateUrl: 'views/directives/edit-environment-from.html'
+  });
+  //function($routeParams, $timeout, $filter, config, utils) {
+  function EditEnvironmentFrom($attrs, $filter, $scope, utils) {
+    var humanizeKind = $filter('humanizeKind');
+    //var canI = $filter('canI');
+    var readOnlySome = [];
+    var cannotDeleteSome = [];
+    //var counter = 1000;
+    //var unique = counter++;
+    //var canIGetSecrets = canI('secrets', 'get');
+    //var canIGetConfigMaps = canI('configmaps', 'get');
+    var ctrl = this;
+
+    ctrl.$id = $scope.$id;
+    //ctrl.entries = ctrl.entries || [{secretRef:{name:''}}];
+    ctrl.setFocusClass = 'edit-environment-from-set-focus-' + ctrl.$id;
+    ctrl.entries = ctrl.entries || [];
+
+    if('cannotAdd' in $attrs) {
+      ctrl.cannotAdd = true;
+    }
+
+    if('cannotDelete' in $attrs) {
+      ctrl.cannotDeleteAny = true;
+    }
+
+    if('cannotSort' in $attrs) {
+      ctrl.cannotSort = true;
+    }
+
+    if('isReadonly' in $attrs) {
+      ctrl.isReadonlyAny = true;
+    }
+
+    if('showHeader' in $attrs) {
+      ctrl.showHeader = true;
+    }
+
+    var addEntry = function(entries, entry) {
+      //entries && entries.push(entry || {secretRef:{name:''}});
+      entries && entries.push(entry || {});
+    };
+
+    ctrl.onAddRow = function() {
+      addEntry(ctrl.entries);
+      utils.setFocusOn('.'+ $scope.setFocusClass);
+    };
+
+    ctrl.deleteEntry = function(start, deleteCount) {
+      ctrl.entries.splice(start, deleteCount);
+      // if the link is used, add a new empty entry to ensure the inputs do not all disappear
+      if(!ctrl.entries.length && ctrl.addRowLink) {
+        addEntry(ctrl.entries);
+      }
+      $scope.forms.editEnvironmentFrom.$setDirty();
+    };
+
+    ctrl.isReadonlySome = function(name) {
+      return _.includes(readOnlySome, name);
+    };
+
+    ctrl.cannotDeleteSome = function(name) {
+      return _.includes(cannotDeleteSome, name);
+    };
+
+    ctrl.isValueFromReadonly = function() {
```

Comment:
We should avoid `valueFrom` in names throughout this control. That refers specifically to choosing an individual key from a secret. It's going to get confusing if we use that in names for `envFrom`.